Add Avalara to tax task

joshuatf opened the pull request:
Fixes #7442 

Adds Avalara as a partner option on the tax task screen.

### Accessibility

<!-- If you've changed or added any interactions, check off the appropriate items below. You can delete any that don't apply. Use this space to elaborate on anything if needed. -->

-   [X] I've tested using only a keyboard (no mouse)
-   [X] I've tested using a screen reader

### Screenshots
![Screen Shot 2021-11-02 at 2 35 13 PM](https://user-images.githubusercontent.com/10561050/139924918-8e4b0d81-1486-428a-948a-41c6e5b2d240.png)

### Detailed test instructions:

Before testing:

**Avalara supported, WooCommerce Tax supported**

1. Select an Avalara and WC Tax supported country (e.g., US) for your store's country
4. Visit the tax task.
4. Make sure your shown the "WooCommerce Tax" and "Avalara" options in the task list

**Avalara not supported, WooCommerce Tax supported**

I don't think this case exists at the moment, but for testing:

1. Remove `US` from the countries in `client/tasks/fills/tax/utils.ts` `supportsAvalara()`
2. Change your store country to the US.
6. Visit the task tax.
7. Make sure you are immediately directed to the WC Tax setup (plugins and connect step should be shown).

**Avalara supported, WooCommerce Tax not supported**

1. Install the TaxJar plugin so that WC Tax is not supported and set your country to an Avalara supported country
6. Visit the task tax.
7. Make sure you are shown only the partner card of Avalara

**Avalara not supported, WooCommerce Tax not supported**

1. Set your store country to one not supported by Avalara or WC Tax (e.g., New Caledonia)
2. Visit the task tax.
7. Make sure you are immediately shown the manual set up flow with the "Configure" button

**Partner actions**
1. Visit the task tax with an Avalara supported country
2. Click Avalara and check that a new tab opens with the WCCOM plugin page
3. Back on the task tax, click on WooCommerce Tax
4. Make sure you are dropped into the old configuration flow (which should be identical to the old flow)

**Events**
1. Enter `localStorage.setItem( 'debug', 'wc-admin:*' );` in your browser's console
2. Set your store's country to an Avalara supported country
3. Note the `wcadmin_tasklist_tax_view_options` event occurs
4. Click on each of the partner action buttons
5. Make sure that `wcadmin_tasklist_tax_select_option` is recorded with the respective `selected_option` partner key.

**Completion**
1. Create a fresh site without ever having set any taxes
2. Note the task is incomplete
3. Install the WC Avalara plugin
4. Check that the task is now marked complete

joshuatf committed:
Move woocommerce tax flow to subdirectory

joshuatf committed:
Add partner cards

joshuatf committed:
Use png for logo files

joshuatf committed:
Add partner card other actions

joshuatf committed:
Add partner card styling

joshuatf committed:
Add in mobile styles

joshuatf committed:
Interpolate links and html elements

joshuatf committed:
Redirect to tax configuration if Avalara not supported

joshuatf committed:
Mark task complete if Alavara is installed

joshuatf committed:
Record events when task partners are shown or selected

joshuatf committed:
Add changelog entry

joelclimbsthings commented on client/tasks/fills/tax/index.tsx:
```diff
@@ -40,64 +32,22 @@ const Tax = ( { onComplete, query } ) => {
+		}
+	);
 
 	const onManual = async () => {
```
You  might wrap this, as well as `onAutomate()` and `onDisable()` in `useCallback`, as any components that are fed these as props will currently always rerender since you're instantiating new functions every time.

joshuatf replied:
Good call, @joelclimbsthings!

joelclimbsthings commented:
Thanks @joshuatf , looks great overall. 

Quick question; if store country is Avalara supported, but not supported by WooCommerce Tax, we still see the option for WooCommerce Tax:
![image](https://user-images.githubusercontent.com/444632/139964119-baa3d66b-0d25-4ea4-81ea-bcad7c5953d1.png)

Is that expected/desired?

Also, my heading style is a little different than expected via your screenshot:

![image](https://user-images.githubusercontent.com/444632/139964277-9dece42e-f13a-4078-b018-13970d60cf3f.png)

Walked through and tested tracks/Avalara install, and all the things. Everything else checkouts out 👍🏻 .

joshuatf committed:
Update task card flow based on visible partners

joshuatf committed:
Skip plugin step if all plugins installed

joshuatf committed:
Move reusable components into subdirectory

joshuatf committed:
Record available partner options

joshuatf committed:
Fix type reference

joshuatf committed:
Wrap callback functions in useCallback to avoid rerenders

louwie17 commented on client/tasks/fills/tax/index.tsx:
```diff
@@ -162,40 +138,117 @@ const Tax = ( { onComplete, query } ) => {
+			return;
 		}
+
+		recordEvent( 'wcadmin_tasklist_tax_view_options', {
```
`wcadmin_` is not necessary here.

joshuatf replied:
Good catch!

louwie17 commented on client/tasks/fills/tax/woocommerce-tax/card.tsx:
```diff
@@ -0,0 +1,79 @@
+			} ) }
+			actionText={ __( 'Continue setup', 'woocommerce-admin' ) }
+			onClick={ () => {
+				recordEvent( 'wcadmin_tasklist_tax_select_option', {
```
`wcadmin_` is not necessary here

joshuatf replied:
👍

louwie17 reviewed (commented):
All the new flows tested well, also very nice work on the structure of your implementation, I quite like how you split every thing up so well kudos!!

I did notice a couple small things, the `recordEvent` calls still had `wcadmin` prepended, which isn't necessary.
I also see the same small "Choose a tax partner" title that Joel had mentioned above.

Lastly, the tax task succeeded fine upon Avalara installation, but clicking on the task still shows the `Enable & set up` button, should we change it to `Continue setup` or `Manage` and redirect to the Avalara settings screen (`/admin.php?page=wc-settings&tab=tax&section=avatax`)?
cc: @pmcpinto on this last question.

pmcpinto commented:
> Lastly, the tax task succeeded fine upon Avalara installation, but clicking on the task still shows the Enable & set up button, should we change it to Continue setup or Manage and redirect to the Avalara settings screen (/admin.php?page=wc-settings&tab=tax&section=avatax)?

👍 changing the copy makes sense. @louwie17 just curious, where did you install the Avalara extension? woocommerce.com or their website?

louwie17 commented:
> @louwie17 just curious, where did you install the Avalara extension? woocommerce.com or their website?

@pmcpinto for testing I downloaded the package from woocommerce.com under my-account/downloads (assuming this is the same package a user gets after they check out) and installed it on my local instance given it isn't available on `.org`.
Is there another way a user could install it?

joshuatf committed:
Handle PR feedback

joshuatf commented:
Agreed on all your comments and thanks for the feedback, @louwie17.  

I was not able to see the card header text in a different size after testing with and without a few versions of GB, but I added explicit font size there so hopefully this is resolved on your end.

I've added a prop to `taskStatus` in the meantime for checking Avalara activation, but this will be migrated to the `additionalData` property in the task list rest api later.

louwie17 commented on client/tasks/fills/tax/index.tsx:
```diff
@@ -162,40 +138,118 @@ const Tax = ( { onComplete, query } ) => {
+	if ( currentPartner ) {
+		return (
+			<TaskCard>
+				{ createElement( currentPartner.component, childProps ) }
```
Could you add a `key` to the props here, it shows a warning for this:
```
{
  key: partner.id,
  ...childProps,
}
```

louwie17 reviewed (approved):
Thanks for the updates, things are testing very well, and the header is showing up correctly now. I did leave one comment to fix a console warning, given it's minor I am approving either way.
Nice work!

joshuatf committed:
Add key to partner card container

joshuatf committed:
Add testing instructions

pmcpinto commented:
> @pmcpinto for testing I downloaded the package from woocommerce.com under my-account/downloads (assuming this is the same package a user gets after they check out) and installed it on my local instance given it isn't available on .org.
> Is there another way a user could install it?

👍 this is the best way to install it. AFAIK they plan to release it on .org, but probably it will only happen next year.
